src/include/compat.h

```diff
@@ -237,12 +240,12 @@ typedef unsigned int uint;
 
 typedef _sigset_t sigset_t;
 
-typedef int uid_t;
-typedef int gid_t;
+typedef unsigned int uid_t;
+typedef unsigned int gid_t;
 
-typedef long blksize_t;
-typedef long blkcnt_t;
-typedef long nlink_t;
+typedef unsigned int blksize_t;
+typedef unsigned __int64 blkcnt_t;
+typedef unsigned short nlink_t;
```

Comment:
One thing that concerns me about this is that the MDS is unfortunately dependent on the types defined for UNIX (Linux). I'm worried that the encoding/decoding routines will not match up for a Windows MDS and Linux MDS.

Rather than redefine these, I would prefer to see the MDS/Client fixed so that the types are no longer dependent on the system types. i.e. define new CephFS types for uid/gid/inodeno_t/etc. in `src/include/fs_types.h`.

Reply:
Thanks for feedback! I did a bit of digging and I think it should be ok for now.

``blksize_t``, ``blkcnt_t`` and ``nlink_t`` are only referenced by ``struct statlite`` https://github.com/ceph/ceph/blob/v16.0.0/src/include/statlite.h#L14-L34, which no longer seems to be used since ``ceph_statx`` was added. Side question, can ``statlite``, ``dirent_plus`` and ``dirent_lite`` be removed?

Also, ``stat`` structures aren't passed over the wire directly, the ``Inode`` Ceph structure is used instead, which doesn't contain any platform specific types other than uid_t and gid_t: https://github.com/ceph/ceph/blob/v16.0.0/src/client/Client.cc#L7201-L7210
https://github.com/ceph/ceph/blob/v16.0.0/src/client/Inode.h#L118-L164

I guess you were suggesting adding ``uid_t`` and ``git_t`` wrappers similar to the already existing ``inodeno_t``, but I'm not sure if we really need that for now, since we're defining those types here, using the Linux definitions as reference.
https://github.com/ceph/ceph/blob/v16.0.0/src/include/fs_types.h#L11-L28

Reply:
> Side question, can statlite, dirent_plus and dirent_lite be removed?

probably? `statlite` was from an old POSIX enhancement proposal (outside of Ceph) that never got any traction.

Reply:
MDS does use `uid_t`/`gid_t`/`inodeno_t`: https://github.com/ceph/ceph/blob/6c805c640527f9600fe04369ac4a5d8ec8583707/src/mds/mdstypes.h#L570-L580

My concern is not so much the over-the-wire encoding but the MDS running on Windows with incompatible types.

Reply:
> My concern is not so much the over-the-wire encoding between MDS-client but the MDS running on Windows with incompatible types that get stored in RADOS.

Got it. Well, we aren't planning on running MDS itself on Windows any time soon, that would probably require quite a few more changes. We're mainly interested in the client side.

Reply:
If you add a guard to abort the MDS if it's running on Windows, I'll be okay with this changeset.

Reply:
Thanks for the suggestion, I've added this check: https://github.com/ceph/ceph/pull/34858/commits/e57ea9203bf38110d6a0f8d281fe8a7e60599736

Reply:
Will that commit not prevent building the client too? i.e. there's no separate `WITH_MDS` CMake check. I was thinking more of a runtime check in `MDSDaemon.cc`.

Reply:
We're currently picking the make targets, skipping ``libmds``, which isn't needed by the client: https://github.com/ceph/ceph/pull/38819/files#diff-40bf9dd98099b94a035f5538c700ace547c94c008e3562bd20d85bfd9e5a782cR166-R177. We did that in order not to overcomplicate the cmake files.

I can switch to a runtime check instead though.